versions/draft-3/SPEC.md

```diff
@@ -994,6 +995,40 @@ task docker_test {
 }
 ```
 
+### Hints Section
+
+Hints are specified in a `hints` section using the same key/value format as the `runtime` section:
+
+```wdl
+task test {
+  command {
+    python script.py
+  }
+  runtime {
+    docker: "ubuntu:latest"
+  }
+  hints {
+    memory: "256MB"
+  }
+}
+```
+
+
+Hints allow a WDL author to provide recommendations to the engine when running a task. 
+
+* Unlike `runtime` these fields are not specified to indicate *what* a workflow does, but rather to provide hints to the engine about *how* to run a workflow.
+* Also unlike `runtime`, the `hints` section allows arbitrary keys and values beyond the set defined below to allow platform specific collaboration between WDL author and platform.
+* The only requirement on new `hints` from the spec is that no engine may implement a `hint` which materially affects the outcome of the task. These should instead be handled as spec changes for additions to the `runtime` section.
```

Comment:
I wonder how difficult this will be to enforce, and what specifically would be considered a material change? IE if I create a file should the MD5 sums match regardless of what the hints section reads? The hints seems like a logical place to put Provider specific information, IE using GPU on Google, but these might actually change the output to some marginal degree since they could be running a modified pipeline.

Reply:
These are some good points and they don't have easy answers! So these are just my gut reactions:

* Enforcement: To some extent it's always going to require either trust in engine implementers *or* some sort of external validation system that can assert that a given workflow works on multiple engines. Maybe Dockstore can eventually fill that role (or at least record whether a workflow is engine-specific as a sort of public shaming?)

* Material change: This is also unfortunately squishy in a world of stochastic analyses. We struggled with the same thing in Cromwell for call caching so my list of "attributes which should be considered important" here is a transcription of the list of "attributes which affect call caching".